### ISSUE TITLE
libretro: OpenGL support not implemented

### ISSUE THREAD

**Issue by rozhuk-im** [xbmc/xbmc#17743]
### Describe the bug
Here is a clear and concise description of what the problem is:

I'm trying to port Play! - PlayStation 2 emulator: https://github.com/rozhuk-im/game.libretro.play
https://purei.org/
https://github.com/jpd002/Play-

It is compiled (with FreeBSD port system, with some private hacks) and run but Kodi (from master) crash because OpenGL in libretro not implemented.

@lrusak , @a1rwulf, @AlwinEsch , @garbear I see some progress in your git repos, but can some one make PR with this with latest changes?

**Comment by a1rwulf** [comment]
Not a real duplicate of: https://github.com/xbmc/xbmc/issues/15874
But it's basically the same :)
@rozhuk-im Thx for bringing it up again, I'll try to find some motiviation to give it another try.

**Comment by rozhuk-im** [comment]
@a1rwulf  Thanks for response, I see #15874, but start this bug as discussion/chat with all who work with it.

At this time I stick with:
``Error: Missing GL version```
after:
```
	glewExperimental = GL_TRUE;
	auto result = glewInit();
```
This is in 
```
	g_hw_render.context_reset = retro_context_reset;
	g_hw_render.context_destroy = retro_context_destroy;
...
	g_environ_cb(RETRO_ENVIRONMENT_SET_HW_RENDER, &g_hw_render);
```
```retro_context_reset()``` function.

**Comment by a1rwulf** [comment]
There's more to it.
I need to verify @garbear's work in the `game.libretro` core library and fix the remaining issues we have inside Kodi.

**Comment by montvid** [comment]
Would be nice to have opengl and vulkan support - kodi retroplayer is a good idea!

**Comment by AlwinEsch** [comment]
About Vulkan, that could for me really a perfect thing for a GSOC project. This has a so big improvement against GL, specially to have App/Process Independent use about them and a Zerocopy inside GPU.

By them can be easily (about his supported parts) a sandbox system added where the parts can use Kodi's GUI in perfect performance and runs safe.

Has looked to it some times ago about the webbrowser thing (hope to start again soon on them), but with a big lack of experience in Vulkan no chance to make byself (only becomed Kodi partly working with a Vulkan <-> GLES emulation library).

Only hard thing then that Kodi must be use them alone without any GL support itself, maybe DX can be still used, but should then also maybe makes sense to change. To allow GLES (e.g. for addons) are some wrapper libs available on Web.

**Comment by KOPRajs** [comment]
> There's more to it.
> I need to verify @garbear's work in the `game.libretro` core library and fix the remaining issues we have inside Kodi.

Is there anything going on with the GL support for Retroplayer? Is it going to work also for Kodi running on GL ES systems (like in LibreELEC on ARM)?

**Comment by lrusak** [comment]
> > There's more to it.
> > I need to verify @garbear's work in the `game.libretro` core library and fix the remaining issues we have inside Kodi.
> 
> Is there anything going on with the GL support for Retroplayer? Is it going to work also for Kodi running on GL ES systems (like in LibreELEC on ARM)?

I haven't worked on it in a while but yes when it's finished GLES will work also.

**Comment by KOPRajs** [comment]
Anything we can do to help or to increase motivation (or donate) to complete this before v20? There is a number of interesting cores relying on the GL support (Muphen, Dolphin, Play!, OpenLara etc.).

**Comment by defcon8** [comment]
Any update?

**Comment by lrusak** [comment]
> Any update?

I'm working on it. I hope to have it for v21 but no promises.

![image](https://user-images.githubusercontent.com/1616460/209459766-7e853ccf-cd2f-422e-aba8-a8ed556961a5.jpeg)

**Comment by KOPRajs** [comment]
Sorry to bother you with the question here, but can you briefly explain how this is supposed to work?

I've just found out that the libretro cores themselves expose support for a different APIs (DX, GL, GLES, or even Vulkan).

So do I get it right, that the Retroplayer needs to implement context for OpenGL, but also for DirectX on Windows and GLES on ARM? And even then, if I run Kodi on GLES platform, I'm only going to run the emulators that has GLES support iself? If the core only supports full OpenGL then there is no luck for ARM boxes with that core, right?

Also I suppose that you are working on the OpenGL context only for now, or is there a chance to get GLES as well in the Omega?

Anyway, thank you for the great work!

**Comment by a1rwulf** [comment]
I think we'll be able to have OpenGL and OpenGL ES support, afaik @lrusak uses Wayland and OpenGL ES for his development.
If a core only supports OpenGL, you'll only be able to run it on OpenGL capable devices ofc.
Note that the mentioned work does not include DX or Vulkan.

**Comment by KOPRajs** [comment]
Thank you for the info. I've just checked and it seems that most of the cores actually should work on GLES which is great!

Funny fact: We actually might temporarily end-up with 3D cores only running on the OpenGL while the shaders are only working for Direct 3D.

**Comment by KOPRajs** [comment]
@lrusak Are you still aiming for Omega now with the alpha1 is released? Is there anything new?

**Comment by lrusak** [comment]
> @lrusak Are you still aiming for Omega now with the alpha1 is released? Is there anything new?

Yes I am. The OpenGL work depends on #22211 which has been a bit of a can of worms and requires some updates for other add-ons. I'll hopefully get around to finishing it soon.

**Comment by KOPRajs** [comment]
What is the current status of the VAOs PR? Does it work? Is it on hold beacuse the OpenGL ES part is missing? Anything we can help with?

**Comment by KOPRajs** [comment]
@lrusak Is there a public repository with the WIP code of the GL support? The https://github.com/xbmc/xbmc/pull/22211 looks good, but it seems to be stuck somehow. What is the showstopper there? I was really hoping to see this happen for Omega.

**Comment by garbear** [comment]
@KOPRajs I've recovered the OpenGL support code and fully rebased it on v21 Beta 2: https://github.com/garbear/xbmc/commits/retro-gl-gbm2

**Comment by KOPRajs** [comment]
@garbear This is really great to hear! I was actually just thinking of trying something similar (to try to get some of the existing code temporarily working, until someone has time to finish it, so it can be merged). I believe @lrusak is the last one who was working on this and he seemed to be pretty close to finishing it. So I was thinking about looking at his code from 2022 (https://github.com/lrusak/xbmc/commits/retro-gl-v3/). This is probably the latest existing code regarding the GL support. However, I'm definitely not competent enough to finish this, so I'm still hoping he will find some time and extra motivation to finish this and he will bring this feature to the level so it can be merged upstream.

**Comment by garbear** [comment]
Small update, I was going through old issues, and rebased the 2020 code on top of Omega v21.1: https://github.com/garbear/xbmc/commits/retro-gl-v3

**Comment by montvid** [comment]
I hope some devs are looking how to merge this. :)

**Comment by garbear** [comment]
None are at the moment sadly. But I'll see if I can upstream some of the plumbing work so that there's less effort needed in the future to push this over the goalpost.

**Comment by KOPRajs** [comment]
> I hope some devs are looking how to merge this into upstream. :)

Unfortunately the code in its current state is not finished and it is not working. I'm also hoping someone can finish this. The biggest issue seems to be the synchronization between the game and the RP regarding when the GPU actually render the content.

**Comment by garbear** [comment]
Builds are uploading soon. I've fully rebased the FBO work on the PR: https://github.com/garbear/xbmc/commits/retro-gl-v3

**Comment by garbear** [comment]
The plumbing PR, https://github.com/xbmc/xbmc/pull/26091, is now code-complete.

I want to write a roadmap outlining the remaining work and challenges. Here's what I got so far:

<hr>

# OpenGL Roadmap for RetroPlayer

OpenGL support in RetroPlayer will enable hardware rendering for cores that require OpenGL. This roadmap outlines the key tasks needed to complete its integration in Kodi.

The OpenGL tracking issue is: https://github.com/xbmc/xbmc/issues/17743

## Current Status

The foundational work for OpenGL support in RetroPlayer has been merged, including:

- **Plumbing APIs:** RetroPlayer APIs now support OpenGL procedures and framebuffer streams ([PR #26091](https://github.com/xbmc/xbmc/pull/26091))
- **Integration with Game API:** The Game API has been extended to connect libretro cores with OpenGL rendering functionality.
- **Core Compatibility:** Initial testing has verified that OpenGL cores, such as Mupen64Plus and Beetle PSX HW, can interface with RetroPlayer without crashing.

## Remaining Work

### VAOs and Rendering Pipeline

The use of Vertex Array Objects (VAOs) is a critical step in modernizing RetroPlayer's rendering pipeline. VAOs streamline the management of vertex attributes, making the OpenGL rendering process more efficient and less error-prone.

- **Current Progress:** The removal of global VAOs has been addressed in [PR #22211: Remove Global VAOs](https://github.com/xbmc/xbmc/pull/22211), which simplifies the rendering pipeline.
- **Next Step:** Review and merge [PR #22211](https://github.com/xbmc/xbmc/pull/22211) to finalize VAO updates.
- The PR has been rebased here: https://github.com/garbear/xbmc/commits/retro-gl-v3

Once the VAO work is completed, RetroPlayer will benefit from a cleaner and more maintainable rendering architecture.

### Frame Buffer Object (FBO) Renderbuffers

Mostly done. Original work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3) and rebased work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3).

<hr>

@KOPRajs Can you explain what we need to do for synchronization?

**Comment by garbear** [comment]
https://github.com/xbmc/xbmc/pull/26091 is now code-complete. I want to include a roadmap outlining the remaining work. Here's what I've generated so far. Let me know if it needs anything else.

# OpenGL Roadmap for RetroPlayer

OpenGL support in RetroPlayer will enable hardware rendering for cores that require OpenGL. This roadmap outlines the key tasks needed to complete its integration in Kodi.

The OpenGL tracking issue is: https://github.com/xbmc/xbmc/issues/17743

## Current Status

The foundational work for OpenGL support in RetroPlayer has been merged, including:

- **Plumbing APIs:** RetroPlayer APIs now support OpenGL procedures and framebuffer streams ([PR #26091](https://github.com/xbmc/xbmc/pull/26091))
- **Integration with Game API:** The Game API has been extended to connect libretro cores with OpenGL rendering functionality.
- **Core Compatibility:** Initial testing has verified that OpenGL cores, such as Mupen64Plus and Beetle PSX HW, can interface with RetroPlayer without crashing.

## Remaining Work

### VAOs and Rendering Pipeline

The use of Vertex Array Objects (VAOs) is a critical step in modernizing RetroPlayer's rendering pipeline. VAOs streamline the management of vertex attributes, making the OpenGL rendering process more efficient and less error-prone.

- **Current Progress:** The removal of global VAOs has been addressed in [PR #22211: Remove Global VAOs](https://github.com/xbmc/xbmc/pull/22211), which simplifies the rendering pipeline.
- **Next Step:** Review and merge [PR #22211](https://github.com/xbmc/xbmc/pull/22211) to finalize VAO updates.
- The PR has been rebased here: https://github.com/garbear/xbmc/commits/retro-gl-v3

Once the VAO work is completed, RetroPlayer will benefit from a cleaner and more maintainable rendering architecture.

### Frame Buffer Object (FBO) Renderbuffers

Mostly done. Original work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3) and rebased work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3).

### Synchronization

The main synchronization challenges for OpenGL support in RetroPlayer stem from the separation of the game loop (which updates game state) and the rendering thread (which processes graphics).

- **GL Context Management:**
  OpenGL contexts are bound to threads, requiring careful management to ensure that rendering operations occur exclusively on  the rendering thread. Context sharing between threads needs to be avoided or strictly synchronized.

- **Dynamic Resolution Changes:**
  Dynamic resolution adjustments during gameplay add complexity. Both threads must coordinate to resize and reallocate frame buffers seamlessly, without causing crashes or visual glitches.

#### GL Context Management

Managing OpenGL contexts in RetroPlayer presents unique challenges due to its multithreaded architecture. Key considerations include:

- **Thread Binding:**
  OpenGL contexts are thread-specific. RetroPlayer must ensure that rendering occurs on a dedicated rendering thread without interfering with the game loop.

- **Context Resetting:**
  When switching contexts (e.g., during resolution changes), RetroPlayer must reinitialize OpenGL state while avoiding disruptions to ongoing operations.

- **Multiple Platforms:**
  Context management strategies must accommodate platform-specific OpenGL implementations, such as OpenGL ES for ARM-based devices or Metal wrappers for macOS.

To address these challenges, RetroPlayer can leverage context creation and management best practices, such as ensuring all rendering occurs on the same thread and using thread-safe proxy functions for OpenGL operations initiated by the game loop.

#### Dynamic Resolution Changes

Needed to accommodate gameplay scenarios where resolution changes are required mid-session, such as switching between fullscreen and windowed modes or adapting to performance constraints. This requires:

- **Real-Time Buffer Resizing:**
  Ensure frame buffers are resized and reallocated without disrupting gameplay or introducing visual artifacts.

- **Synchronization with Rendering Thread:**
  Both the game loop and the rendering thread must coordinate to apply resolution changes seamlessly.

---

## Conclusion

RetroPlayer's OpenGL support is well underway, with foundational work completed and key APIs in place. The remaining tasks include finalizing VAO updates, completing FBO renderbuffers, addressing synchronization challenges, and implementing dynamic resolution support. With these enhancements, RetroPlayer will achieve robust OpenGL integration, paving the way for future advancements like Vulkan support and broader hardware compatibility.

**Comment by KOPRajs** [comment]
I'll try to add what I've got so far. Please note that I have zero experience in this and I might be off in details.

First of all, if I get it right, calling an OpenGL function is not a direct command for the GPU. As there may be more threads doing the rendering via OpenGL, or even more applications, there is basically a queue of OpenGL commands managed by the GL driver. The GL commands from different threads are not guaranteed to be executed in the same order as they were called.

We've got 2 rendering threads:
1. The libretro core (the emulator itself) which is rendering to FBO (basically into a texture).
2. The libretro frontend (RetroPlayer) which is responsible for stretching the texture to the screen, appling shaders on it and presenting it to the user.

This is how libretro API is designed and there is no way how we can do OpenGL rendering in one thread.

The current code seems almost working (I've got some frames rendered correctly, but it is flickering).

So this is the issue:
1. RetroPlayer creates a FBO (a framebuffer with a texture attached to it).
2. RetroPlayer calls the libretro core requesting it to render next frame into this FBO.
3. The libretro core calls it's OpenGL functions to draw the frame, but they are not executed by the GPU immediately.
4. The libretro core tells RetroPlayer that it has finished.
5. RetroPlayer calls it's own sequnce of OpenGL commands which draw the texture from the FBO to the screen.

Now 2 things can happen based on the timing and the GL driver:
- The commands from the libretro core's thread have been already executed and the frame is ready.
- Or the frame haven't been rendered to the texture by the GPU driver yet, so the texture is black.

There are several options how we can prevent this and I don't fully understand benefits of each of them, but most promising seems to be the use of the OpenGL Sync objects called Fences (described here: https://arm-software.github.io/opengl-es-sdk-for-android/thread_sync.html).

When browsing through the RetroArch's code I've found out that it is using it as well: https://github.com/libretro/RetroArch/blob/739f2ff55ce9f2b31eade0c503c7f4e5578037a5/gfx/drivers/gl2.c#L2410

**Comment by KOPRajs** [comment]
@garbear I've tried to get the current version of the WIP code to work with the already merged part, but so far I struggle to get it to work the same as before. I've noticed, that you've done some changes and cleanups before the merge, but I'm possibly missing something.

Can you please have a look and help me to get it to work the same as before?

I've commited the changes I've done so far as the [WIP] on top of the code based on retroplayer-21.2 in my branch:
https://github.com/KOPRajs/xbmc/commits/retroplayer-gl-v4/

I've restored the 2 removed methods in _CRPRenderManager_ and uncommented the return value in _OpenStream_ and _EnableHardwareRendering_.

I'm getting sound with black screen for Mupen64 and no errors in log.

I expect that the changes done here might be the cause:
https://github.com/xbmc/xbmc/compare/50931a4cd2b499c601b3a2653a68961197f1c1a5..58c4557dc431daaf4c37ad4161c351315f24f26c

The _m_renderManager.Configure()_ is maybe missing correct pixel format, but since you've removed the attributes from _StreamProperties_, I don't know, where do I get them.

**Comment by mirko** [comment]
Is there any news on this? Wondered if it got upstreamed meanwhile, however it seems OpenGL contexts still can't be opened from within Kodi.
Slightly off-topic: is it the same or a totally different thing with Vulkan-backed retroarch-cores (as many can be configured using OpenGL *or* Vulkan).

**Comment by KOPRajs** [comment]
Unfortunately, there is nothing new here. Some parts of the code were merged upstream, but the rest is not being developed at all. I'm not aware of anyone currently working on this. I would love to see this myself, but currently fixing remaining bugs in the new video shader support before v22 release is on top of the TODO.

The Vulkan is even more far away. The Kodi itself would have to start using Vulkan first.

**Comment by mirko** [comment]
Ack, thanks for the update. I don't quite see/understand why libretro cores can't just use the vulkan API to their liking -  for OpenGL I figured it's because kodi is already opening and maintaining a context, but maybe I'm completely mistaken? Either way, a bummer work on this is stalled for the moment, but who am I to tell FOSS-devs (or ppl in general) what to focus on :)

**Comment by KOPRajs** [comment]
The libretro cores can't manage their graphics context, they share the context with the frontend app, which is Kodi in this case. There is currently no Vulkan context to share with the libretro cores in Kodi.

Also, if I understand it correctly, even if we have support for multiple accelerated libretro drivers one day, you won't be able to switch between them in Kodi, like you can in RetroArch.
If you run Kodi on Windows using the Direct3D renderer, you are going to be limited to Direct3D context support for libretro cores.
If you run Kodi in Linux using the OpenGL, you are going to be limited to use OpenGL renderer in libretro cores.
If we ever get Kodi running in Vulkan, you are going to be limited to Vulkan renderers only for the libretro cores.

See how RetroArch is closing its window completely and opens a new one when starting a game. That's where it can switch the graphics API. I don't expect this is going to be ever possible in Kodi.
In Kodi, the supported graphics API is being chosen in compile time.

With that being said, it would still be great if we can get the OpenGL (ES) support, which is the closest one to happen. Unfortunately, there is currently no one having both the needed skills and a free time for this.

**Comment by garbear** [comment]
OpenGL support arrives, in experimental form! https://github.com/garbear/xbmc/pull/151

**Comment by KOPRajs** [comment]
Maybe we can link this issue to be finally closed when #29303 is merged upstream?
